src/Campaigns/Routes/GetCampaigns.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Give\Campaigns\Routes;
+
+use Exception;
+use Give\API\RestRoute;
+use Give\Campaigns\Models\Campaign;
+use Give\Campaigns\Repositories\CampaignRepository;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * @unreleased
+ */
+class GetCampaigns implements RestRoute
+{
+
+    /** @var string */
+    protected $endpoint = 'campaigns';
+
+    /**
+     * @unreleased
+     */
+    public function registerRoute()
+    {
+        register_rest_route(
+            'give-api/v2',
+            $this->endpoint,
+            [
+                [
+                    'methods' => WP_REST_Server::READABLE,
+                    'callback' => [$this, 'handleRequest'],
+                ],
+                'args' => [
+                    'page' => [
+                        'type' => 'integer',
+                        'default' => 1,
+                        'minimum' => 1,
+                    ],
+                    'per_page' => [
+                        'type' => 'integer',
+                        'default' => 30,
+                        'minimum' => 1,
+                    ],
+                ],
+            ]
+        );
+    }
+
+    /**
+     * @unreleased
+     *
+     * @throws Exception
+     */
+    public function handleRequest(WP_REST_Request $request): WP_Rest_Response
+    {
+        $page = $request->get_param('page');
+        $perPage = $request->get_param('per_page');
+
+        $query = give(CampaignRepository::class)->prepareQuery();
+
+        $query
+            ->limit($perPage)
+            ->offset(($page - 1) * $perPage);
+
+        $campaigns = $query->getAll() ?? [];
+        $totalCampaigns = empty($campaigns) ? 0 : Campaign::query()->count();
+        $totalPages = (int)ceil($totalCampaigns / $perPage);
+
+        $response = rest_ensure_response($campaigns);
+        $response->header('X-WP-Total', $totalCampaigns);
+        $response->header('X-WP-TotalPages', $totalPages);
```

Comment:
I'd recommend setting response header links for pages. That way it works with things like the `@wordpress/api-fetch` fetch all middleware.

Eg: https://github.com/ithemes/ithemes-security-pro/blob/73caf6c88c69a4a3ecbf78f4565423ebf36be7ed/core/lib/class-itsec-lib-rest.php#L334

Reply:
I don't have access to this repo.

Reply:
```

public static function paginate( WP_REST_Request $request, WP_REST_Response $response, int $count, string $path ) {
		$max_pages = ceil( $count / $request['per_page'] );
		$response->header( 'X-WP-Total', $count );
		$response->header( 'X-WP-TotalPages', $max_pages );

		$request_params = $request->get_query_params();
		$base           = add_query_arg(
			map_deep( $request_params, function ( $value ) {
				if ( is_bool( $value ) ) {
					$value = $value ? 'true' : 'false';
				}

				return urlencode( $value );
			} ),
			rest_url( $path )
		);

		if ( $request['page'] > 1 ) {
			$prev_page = $request['page'] - 1;

			if ( $prev_page > $max_pages ) {
				$prev_page = $max_pages;
			}

			$prev_link = add_query_arg( 'page', $prev_page, $base );
			$response->link_header( 'prev', $prev_link );
		}

		if ( $max_pages > $request['page'] ) {
			$next_page = $request['page'] + 1;
			$next_link = add_query_arg( 'page', $next_page, $base );

			$response->link_header( 'next', $next_link );
		}
	}

```

Reply:
Resolved:  https://github.com/impress-org/givewp/pull/7547/commits/7c3e40fb01c5b7071eb5eedfef4909fa168f59f6

Reply:
@alaca You have access to that repo now, FYI.